src/uct/tcp/tcp_ep.c

```diff
@@ -62,64 +105,207 @@ static void uct_tcp_ep_epoll_ctl(uct_tcp_ep_t *ep, int op)
 
 static inline int uct_tcp_ep_can_send(uct_tcp_ep_t *ep)
 {
-    ucs_assert(ep->offset <= ep->length);
+    ucs_assert(ep->tx->offset <= ep->tx->length);
     /* TODO optimize to allow partial sends/message coalescing */
-    return ep->length == 0;
+    return ep->tx->length == 0;
+}
+
+static ucs_status_t uct_tcp_ep_ctx_init(uct_tcp_iface_t *iface,
+                                        uct_tcp_ep_ctx_t **ctx)
+{
+    ucs_status_t status;
+
+    *ctx = ucs_malloc(iface->config.buf_size, "tcp_ctx");
+    if (*ctx == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    (*ctx)->buf = ucs_malloc(iface->config.buf_size, "tcp_buf");
+    if ((*ctx)->buf == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_free_ctx;
+    }
+
+    (*ctx)->offset = 0;
+    (*ctx)->length = 0;
+
+    return UCS_OK;
+
+err_free_ctx:
+    ucs_free(*ctx);
+    *ctx = NULL;
+    return status;
+}
+
+static void uct_tcp_ep_ctx_cleanup(uct_tcp_ep_ctx_t **ctx)
+{
+    ucs_free((*ctx)->buf);
+    (*ctx)->buf = NULL;
+
+    ucs_free(*ctx);
+    *ctx = NULL;
+}
+
+UCT_TCP_EP_DEFINE_CTX_OPS(tx)
+UCT_TCP_EP_DEFINE_CTX_OPS(rx)
+
+static void uct_tcp_ep_addr_cleanup(ucs_sock_addr_t *sock_addr)
+{
+    if (sock_addr == NULL){
+        return;
+    }
+
+    if (sock_addr->addr != NULL) {
+        ucs_free((void*)sock_addr->addr);
+    }
+
+    memset(sock_addr, 0, sizeof(*sock_addr));
+}
+
+static ucs_status_t uct_tcp_ep_addr_init(ucs_sock_addr_t *sock_addr,
+                                         const struct sockaddr *addr)
+{
+    socklen_t addr_len = sizeof(struct sockaddr_in); /* TODO */
+    struct sockaddr *new_addr;
+
+    if (sock_addr == NULL) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if (addr == NULL) {
+        uct_tcp_ep_addr_cleanup(sock_addr);
+    } else {
+        new_addr = ucs_malloc(addr_len, "sock_addr");
+        if (new_addr == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+        sock_addr->addr    = new_addr;
+        sock_addr->addrlen = addr_len;
+    }
+
+    return UCS_OK;
+}
+
+static void uct_tcp_ep_close_fd(int *fd_p)
+{
+    if ((fd_p != NULL) && (*fd_p != -1)) {
+        close(*fd_p);
+    }
+}
+
+static unsigned uct_tcp_ep_in_iface(uct_tcp_ep_t *ep)
+{
+    return ucs_list_is_empty(&ep->list) == 0;
+}
+
+static void uct_tcp_ep_del_from_iface(uct_tcp_iface_t *iface,
+                                      uct_tcp_ep_t *ep)
+{
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    if (uct_tcp_ep_in_iface(ep) == 1) {
+        ucs_list_del(&ep->list);
```

Comment:
do we need this check? ep is added always (also it looks safe to call del anyway)

Reply:
sure